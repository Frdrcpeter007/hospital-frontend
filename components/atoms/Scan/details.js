import React, { useContext, useEffect, useState } from "react";
import { useSession } from "next-auth/react";
import Fetcher from '../../../lib/Fetcher';
import { BiTransferAlt } from 'react-icons/bi';
import CopyToClipboard from 'react-copy-to-clipboard';
import { AiOutlineExclamationCircle } from 'react-icons/ai';
import { MdPayments } from 'react-icons/md';
import { StepContext } from ".";
import { useRouter } from "next/router";
import { convertCurrency } from "../../../lib/helper";


const ScanDetails = ({ shorten }) => {

	const { step, setStep, total } = useContext(StepContext);
	const router = useRouter();

	const { data: session } = useSession();
	const [copy, setCopy] = useState(false);
	const [data, setData] = useState(null);
	const [dataAmount, setDataAmount] = useState(null);
	const [dataAmountHospital, setDataAmountHospital] = useState({ amount: total, currency: "CDF" });
	const [currencyProvider, setCurrencyProvider] = useState('CDF');
	const [convertRequest, setConvertRequest] = useState(false);

	const Options = {
		method: "GET",
		headers: {
			'Content-Type': 'application/json',
			'Authorization': `Bearer ${session.user.data.access_token}`
		}
	}

	useEffect(() => {
		fetch(`https://api.wiiqare-app.com/api/v1/provider/provider-voucher-details?shortenHash=${shorten}`, Options).then(async res => {
			let json = await res.json();

			if (/AUTH_TOKEN_EXPIRED/i.test(json.code)) router.push('/login')

			setData(json)
			setDataAmount({ amount: json.amount, currency: json.currency })

			setConvertRequest(true);

			const response = await convertCurrency(
				json.currency,
				json.amount,
				'CDF'
			);

			setConvertRequest(false);
			setDataAmount({ amount: response.result, currency: response.query.to })
		});
	}, []);

	const handleCurrency = async (e) => {
		setCurrencyProvider(e.target.checked == true ? 'USD' : 'CDF');

		setConvertRequest(true);
		const res = await convertCurrency(
			dataAmount.currency,
			dataAmount.amount,
			e.target.checked == true ? 'USD' : 'CDF',
		);

		const resHospital = await convertCurrency(
			"CDF",
			total,
			e.target.checked == true ? 'USD' : 'CDF',
		);

		setConvertRequest(false);

		setDataAmount({ amount: res.result, currency: res.query.to })
		setDataAmountHospital({ amount: resHospital.result, currency: resHospital.query.to })
		console.log(resHospital);
	}


	const SliceText = ({ text }) => {
		return <>{text.slice(0, 9)}...{text.slice(-8)}</>
	}

	if (!data) return (
		<div className="flex flex-col gap-3 items-center mt-14 h-full">
			<div className="animate-spin inline-block w-4 h-4 border-[2px] border-current border-t-transparent text-gray-700 rounded-full" role="status" ariaLabel="loading">
				<span className="sr-only">Loading...</span>
			</div>
		</div>
	)

	if (data.code) return (
		<div className='mx-auto flex justify-center items-center flex-col h-full gap-3'>
			<img src="/images/qr-code-error.png" alt="QR Code erroné" className='w-40 opacity-70' />
			<span className='text-gray-500 text-sm font-light'>{data.description}</span>
			<button onClick={() => location.reload()} className='bg-orange text-white text-sm px-3 py-2 rounded-lg effect-up shadow-sm font-light'>Scannez à nouveau</button>
		</div>
	)

	return (
		<div className='flex justify-center flex-col gap-6 h-full items-center mx-auto py-4 md:py-10 mb-20'>
			<div className='md:w-1/3 sm:w-2/3 w-full bg-white rounded-xl py-8 min-h-fit flex '>
				<div className="flex flex-col gap-6 justify-center w-full">
					<div className="flex flex-col items-center text-center space-y-2">
						<div className="flex flex-col items-center select-none">
							<h1 className="font-extrabold text-gray-700 text-lg hidden md:flex">Détails Pass santé</h1>
						</div>
					</div>

					<div className="flex gap-4 items-center px-5 justify-center">
						<img className="inline-block h-[2.875rem] w-[2.875rem] rounded-full ring-2 ring-white dark:ring-gray-800" src="/images/homme.png" alt="Image Description" />
						<BiTransferAlt size={30} className='text-gray-400' />
						<img className="inline-block h-[2.875rem] w-[2.875rem] rounded-full ring-2 ring-white dark:ring-gray-800" src="https://i.goopics.net/9w9mvb.png" alt="Image Description" />
					</div>


					<span className="text-xs flex justify-center items-center gap-1">Pass Sante ID:
						<CopyToClipboard text={data.hash} onCopy={() => {
							setCopy(true); setTimeout(() => {
								setCopy(false)
							}, 2000);
						}}>
							<div className="flex items-center gap-1">
								[
								<div className="tooltip" data-tip={!copy ? "Copy to clipboard" : "✓ Copy"}>
									<span className="text-orange cursor-pointer"><SliceText text={data.hash} /></span>
								</div>
								]
							</div>
						</CopyToClipboard>
					</span>


					<div className="flex justify-between px-6 items-center">
						<h1 className='font-normal text-md text-gray-400'>Choisir la devise : </h1>

						{convertRequest ? (
							<div className="pr-8">
								<div
									className="animate-spin inline-block w-4 h-4 border-[2px] border-current border-t-transparent text-gray-400 rounded-full"
									role="status"
									ariaLabel="loading"
								>
									<span className="sr-only">Loading...</span>
								</div>
							</div>
						) : (
							<label for="Toggle4" className="inline-flex items-center rounded-full overflow-hidden cursor-pointer bg-white border-2 border-primary text-primary">
								<input id="Toggle4" type="checkbox" className="hidden peer" onChange={(e) => handleCurrency(e)} />
								<span className="px-4 py-2 bg-transparent peer-checked:bg-primary text-primary peer-checked:text-white font-semibold select-none text-sm">CDF</span>
								<span className="px-4 py-2 bg-primary text-white font-semibold peer-checked:bg-white peer-checked:text-primary select-none text-sm">USD</span>
							</label>
						)}


					</div>

					<ItemsDetails title={"Nom du Patient"} value={data.patientNames} exclamation={true} />
					<ItemsDetails title={"Hôpital"} value={session?.user?.data.names ?? session?.user?.data.name} exclamation={false} />
					<ItemsDetails title={"Montant Pass Santé"} value={new Intl.NumberFormat("en-US", { style: 'currency', currency: dataAmount.currency }).format(dataAmount.amount)} otherValue={`${(dataAmount.amount - dataAmountHospital.amount) > 0 ? 'Reste: ' : 'Manquant: '}` + new Intl.NumberFormat("en-US", { style: 'currency', currency: currencyProvider }).format(dataAmount.amount - dataAmountHospital.amount)} />
					<ItemsDetails title={"Montant Hopital"} value={new Intl.NumberFormat("en-US", { style: 'currency', currency: dataAmountHospital.currency }).format(dataAmountHospital.amount)} otherValue={new Intl.DateTimeFormat('fr-FR', { dateStyle: 'medium' }).format(new Date())} />

					{
						total > data.amount ? (
							<span className="text-xs text-red-400 font-normal px-4">* Ce traitement coûte plus cher que le montant dans ce Pass Santé</span>
						) : <></>
					}

					<div className='flex justify-center'>
						<button className='capitalize bg-orange w-fit  px-6 py-4 rounded-xl text-white flex gap-2 items-center effect-up shadow-md' onClick={() => setStep(step + 1)}><MdPayments size={20} /> {total > data.amount ? 'Payer une partie avec ce Pass' : 'Procéder au paiement'}</button>
					</div>

				</div>
			</div>
		</div>
	);
}

export default ScanDetails;

function ItemsDetails({ title, value, otherValue, exclamation = false, last = true }) {
	return (
		<div className="flex flex-col hover:shadow-sm w-full">
			<div className='flex justify-between px-6 py-2 items-center'>
				<div className='flex flex-col  mb-2 gap-1'>
					<h1 className='font-normal text-gray-400 text-sm'>{title}</h1>
					<div className='flex gap-2 items-center'>
						<h3 className='font-bold text-md text-gray-700'>{value}</h3>
						{otherValue && <span className='text-sm text-gray-400'> - {otherValue}</span>}
					</div>
				</div>

				{exclamation && <AiOutlineExclamationCircle size={24} className='text-gray-300 cursor-pointer hover:text-orange transition-all duration-200' />}
			</div>

			{last && <hr />}
		</div>
	)
}