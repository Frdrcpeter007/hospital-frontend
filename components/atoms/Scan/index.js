import MuiPhoneNumber from 'material-ui-phone-number';
import React, { Fragment, useState } from 'react';
import { QrReader } from 'react-qr-reader';
import { Dialog, Transition } from '@headlessui/react';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import Box from '@mui/material/Box';
import PropTypes from 'prop-types';
import Typography from '@mui/material/Typography';
import { Button, Stack, TextField } from '@mui/material';
import { useFormik } from 'formik';
import * as yup from "yup";
import ScanDetails from './details';
import SecurityCode from './security';

const Scan = () => {
	const [data, setData] = useState(null);
	const [step, setStep] = useState(0);
	const [value, setValue] = useState(0);

	const handleChange = (event, newValue) => {
		setValue(newValue);
	};

	const handleStep = (result = null) => {
		
		setStep(step+1)
		result && setData(result && result.slice(0, 8))

	};	

	if (step === 0)
		return (
			<Transition appear show={true} as={"div"} className="flex justify-center flex-col gap-6 h-full items-center mx-auto py-4 md:py-10">
				<Transition.Child
					enter="ease-out duration-300"
					enterFrom="opacity-0 scale-95"
					enterTo="opacity-100 scale-100"
					leave="ease-in duration-200"
					leaveFrom="opacity-100 scale-100"
					leaveTo="opacity-0 scale-95"
					className={"md:w-1/2 bg-white rounded-xl px-10 py-8 min-h-fit"}
				>

					<TabsModal value={value} handleChange={handleChange} />


					<TabItems value={value} handleStep={handleStep} />


				</Transition.Child>
			</Transition>

		)

	if (step === 1) return <ScanDetails shorten={data} handleStep={handleStep} />

	return <SecurityCode />
};

export default Scan;

function TabsModal({ value, handleChange }) {

	return (
		<div>
			<Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
				<Tabs value={value} onChange={handleChange} aria-label="basic tabs example">
					<Tab label={"Scan QR Code"} {...a11yProps(0)} key={0} />
					<Tab label={"Write Code Voucher"} {...a11yProps(1)} key={1} />
				</Tabs>
			</Box>
		</div>
	)
}

function TabItems({ value, handleStep }) {


	const onSubmit = async (values) => {
        if (Object.keys(values).length == 0) return console.log("Pas de données");

        handleStep(1, {transactionHash: "0xf59b12eccfc5faedbc4657bd593d6d6a0c679623"})
    };

	const ValidationSchema = yup.object().shape({
        pass: yup.string().required("Pass Santé est requis")
    });

	const formik = useFormik({
        initialValues: {
            pass: '',
        },
        validationSchema: ValidationSchema,
        onSubmit
    })

	const renderError = (message) => (
        <p className="text-xs text-red-600 font-light flex items-center gap-1 px-1">{message}</p>
    );

	return (
		<div className="mt-2">
			<div className="text-sm text-gray-500">
				{/* For Scan */}
				<TabPanel value={value} index={0} >
					<div className="mt-4">

						<div className='text-center'>
							<h3 className='text-xl font-semibold text-gray-700'>Scan QR Code</h3>
							<p className='text-sm font-light text-gray-500'>Accorder la demande d'allumer la caméra</p>
						</div>
						<div className="">
							<QrReader
								onResult={(result, error) => {
									if (!!result) {
										handleStep(result.text)
									}
								}}
								constraints={{ facingMode: "environment" }}
								style={{ width: '100%' }}
							/>
						</div>
					</div>

				</TabPanel>

				<TabPanel value={value} index={1} >

					<div className="space-y-8 py-10 px-20">
                       
                        <form id="signupform" onSubmit={formik.handleSubmit}>
                            <Stack spacing={2}>

                                <div className="space-y-1">
                                    <TextField
                                        id="outlined-basic"
                                        fullWidth
                                        label="Entrez le code Pass Santé"
                                        variant="outlined"
                                        name="pass"
										{...formik.getFieldProps('pass')}

                                    />
                            		{formik.errors.pass && formik.touched.pass ? renderError(formik.errors.pass) : <></>}
                                </div>

                                <div className="form-button flex flex-row-reverse">
                                    <Button size="medium" variant="contained" type="submit">
										Soumettre
                                    </Button>
                                </div>
                            </Stack>
                        </form>
					</div>

				</TabPanel>
			</div>
		</div>
	)
}

export function TabPanel(props) {
	const { children, value, index, ...other } = props;

	return (
		<div
			role="tabpanel"
			hidden={value !== index}
			id={`simple-tabpanel-${index}`}
			aria-labelledby={`simple-tab-${index}`}
			{...other}
		>
			{value === index && (
				<Box>
					<Typography>{children}</Typography>
				</Box>
			)}
		</div>
	);
}

TabPanel.propTypes = {
	children: PropTypes.node,
	index: PropTypes.number.isRequired,
	value: PropTypes.number.isRequired,
};

function a11yProps(index) {
	return {
		id: `simple-tab-${index}`,
		'aria-controls': `simple-tabpanel-${index}`,
	};
}