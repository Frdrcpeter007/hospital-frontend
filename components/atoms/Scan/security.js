import { Button, TextField } from '@mui/material';
import { useFormik } from 'formik';
import { useSession } from 'next-auth/react';
import React, { useState } from 'react';
import { useMutation } from 'react-query';
import * as yup from "yup";
import { sendSecurityCode } from '../../../lib/helper';
import LoadingButton from '../Loader/LoadingButton';

const SecurityCode = ({shorten}) => {
	const { data:session } = useSession();
	const [state, setState] = useState({type: 0, message: ""});

	console.log(session.user.data.access_token);

	const sendEmailMutation = useMutation(sendSecurityCode,  {
        onSuccess: (res) => {

            if(!res.code) {
                setState({type: 1, message: "Un mail vous a été envoyé"})
                setComplete(true)
            } else {
                setState({type: 2, message:res.message ?? res.description})
                setTimeout(() => {
                    setState({ type: 0, message: "" })
                }, 3000);
            };
        }
    });

	const ValidationSchema = yup.object().shape({
		securityCode: yup.string().required("Code de sécurité est requis").max(6, "6 caractère pour le code de sécurité"),
	});

	const onSubmit = async (values) => {
        if (Object.keys(values).length == 0) return console.log("Pas de données");
        //dispatch(setRegister({...values}))
        sendEmailMutation.mutate({...values, shortenHash: shorten, providerId: session.user.data.userId, accessToken: session.user.data.access_token})
    };

	const formik = useFormik({
        initialValues: {
            securityCode: ''
        },
        validationSchema: ValidationSchema,
        onSubmit
    })

	const renderError = (message) => (
		<p className="text-xs text-red-600 font-light flex items-center gap-1 px-1">{message}</p>
	);

    return (
        <div className='h-full flex justify-center items-center pb-40'>
				<div className="w-full  mx-auto max-w-md p-6 overflow-hidden text-left align-middle transition-all transform bg-white shadow-sm rounded-xl space-y-8 mt-2">

					<h1 className='text-center font-semibold text-xl text-gray-700'>Entrez le code de sécurité</h1>

					<div className='space-y-3'>
						<div className="flex flex-col gap-2 w-full">

							<form className="flex flex-col w-full justify-between items-center gap-3" onSubmit={formik.handleSubmit} id='form-security'>

								<TextField
									fullWidth
									name="securityCode"
									label="Entrez le code de sécurité"
									variant="outlined"
									placeholder={"Entrez le code de sécurité"}
									className="w-full placeholder:text-gray-400 hover:outline-none focus:ring-0 border border-gray-300 rounded-lg focus:ring-sky"
									{...formik.getFieldProps('securityCode')}
								/>
                            	
								{formik.errors.securityCode ?  renderError(formik.errors.securityCode): <></>}

							</form>
						</div>

						<div className="flex flex-row-reverse gap-4">
							<button className="bg-orange text-md py-2 px-4 rounded-lg effect-up text-white" type='submit' form='form-security'>
								{sendEmailMutation.isLoading ? <LoadingButton /> : 'Envoyer'}
							</button>
						</div>
					</div>
				</div>
			</div>
    );
}

export default SecurityCode;
