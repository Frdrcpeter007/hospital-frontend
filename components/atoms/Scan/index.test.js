import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import Scan from '.';
import { SessionProvider } from 'next-auth/react';

describe('Scan', () => {
  let component;
  beforeEach(() => {
    const res = render(
      <SessionProvider
        session={{
          user: { name: 'John Doe', data: { providerId: 'test123' } },
        }}
      >
        <Scan />
      </SessionProvider>,
    );
    component = res.container;
  });

  it('should render the component', () => {
    expect(component).toMatchSnapshot();
  });

  test('should render text elements', () => {
    expect(screen.getByText('DÃ©finir comme traitement')).toBeInTheDocument();
  });
});
