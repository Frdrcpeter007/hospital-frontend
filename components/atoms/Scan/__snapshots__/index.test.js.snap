// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scan should render the component 1`] = `
<div>
  <div
    className="flex justify-center flex-col gap-6 h-full items-center mx-auto py-4 md:py-10"
  >
    <div
      className="md:w-1/2 bg-white rounded-xl px-10 py-8 min-h-fit ease-out duration-300 opacity-0 scale-95"
    >
      <div>
        <div
          className="MuiBox-root css-1gsv261"
        >
          <div
            className="MuiTabs-root css-1ujnqem-MuiTabs-root"
          >
            <div
              className="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
              style="overflow: hidden; margin-bottom: 0px;"
            >
              <div
                aria-label="basic tabs example"
                className="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                role="tablist"
              >
                <button
                  aria-controls="simple-tabpanel-0"
                  aria-selected="true"
                  className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                  id="simple-tab-0"
                  role="tab"
                  tabindex="0"
                  type="button"
                >
                  Scan QR Code
                  <span
                    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-controls="simple-tabpanel-1"
                  aria-selected="false"
                  className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                  id="simple-tab-1"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  Saisir Pass Santé
                  <span
                    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <span
                className="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
                style="left: 0px; width: 0px;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="mt-2"
      >
        <div
          className="text-sm text-gray-500"
        >
          <div
            aria-labelledby="simple-tab-0"
            id="simple-tabpanel-0"
            role="tabpanel"
          >
            <div
              className="MuiBox-root css-0"
            >
              <p
                className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                <div
                  className="mt-4"
                >
                  <div
                    className="text-center"
                  >
                    <h3
                      className="text-xl font-semibold text-gray-700"
                    >
                      Scan QR Code
                    </h3>
                    <p
                      className="text-sm font-light text-gray-500"
                    >
                      Accorder la demande d'allumer la caméra
                    </p>
                  </div>
                  <div
                    className=""
                  >
                    <section>
                      <div
                        style="width: 100%; padding-top: 100%; overflow: hidden; position: relative;"
                      >
                        <video
                          id="video"
                          style="top: 0px; left: 0px; width: 100%; height: 100%; display: block; overflow: hidden; position: absolute;"
                        />
                      </div>
                    </section>
                  </div>
                </div>
              </p>
            </div>
          </div>
          <div
            aria-labelledby="simple-tab-1"
            hidden=""
            id="simple-tabpanel-1"
            role="tabpanel"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
